### PR TITLE
RFC: Use objects instead of symbols to represent platforms

### PR DESCRIPTION
**Pull request by ararslan**
Basic sketch of an approach to using objects and dispatch rather than symbols and dict lookups. Definitely WIP as the design is of course open to suggestions, plus this still needs tests and adjustment to any other code that assumes symbols.

### CODE REVIEW

**Comment by staticfloat** [comment]
It would be great to have `show()` methods for the platforms so that when spitting out `platform => (url, hash)` pairings, I can just print the platform and it's a valid constructor syntax.

**Comment by ararslan** [comment]
How are you envisioning it should print? It should already be printing as a valid constructor; for example, `Linux(:x86_64)` prints as written for me.

**Comment by staticfloat** [comment]
Yes, glibc was the only thing I was worried about. How about this. Lets make the field “libc” and the values “glibc” or “musl”, etc....

**Comment by ararslan** [comment]
Done. `Linux` objects print with both fields: `arch` and `libc`, and a matching 2-arg constructor exists, so printed output can be copy-pasted to make an equivalent object.

Anything else? I cleaned up my mess so tests are passing now.

File: test/runtests.jl [hunk 1]
```diff
@@ -91,43 +91,52 @@ BinaryProvider.probe_platform_engines!(;verbose=true)
...
-    @test platform_is_windows(:win32)
-    @test !platform_is_windows(:linux64)
-    @test platform_is_apple(:mac64)
-    @test !platform_is_apple(:linuxppc64le)
+    @test Compat.Sys.islinux(Linux(:aarch64))
+    @test !Compat.Sys.islinux(Windows(:x86_64))
+    @test Compat.Sys.iswindows(Windows(:i686))
+    @test !Compat.Sys.iswindows(Linux(:x86_64))
+    @test Compat.Sys.isapple(MacOS())
+    @test !Compat.Sys.isapple(Linux(:ppc64le))
```
Comment (staticfloat) [inline test/runtests.jl]: You need to add a `using Compat` at the top of this file.  How does this possibly pass Travis on 0.6?!

Comment (ararslan) [reply test/runtests.jl]: I... don't know.

Comment (staticfloat) [reply test/runtests.jl]: This more than anything convinces me that I have absolutely no idea how computers work.

Comment (ararslan) [reply test/runtests.jl]: Oddly enough, 0.7 caught it but not 0.6.

Comment (ararslan) [reply test/runtests.jl]: And AppVeyor didn't seem to notice

Comment (ararslan) [reply test/runtests.jl]: TwilightZone.jl

**Comment by ararslan** [comment]
Okay, _now_ tests are passing.

File: src/PlatformNames.jl [hunk 2]
```diff
@@ -1,154 +1,193 @@
...
+```
 """
-function platform_is_windows(platform::Symbol)
-    const easily_shattered = [
-        :win32,
-        :win64,
-    ]
-    return platform in easily_shattered
-end
+triplet(l::Linux) = arch(l) === :armv7l ? "arm-linux-gnueabihf" : string(arch(l), "-linux-gnu")
```
Comment (tkelman) [inline src/PlatformNames.jl]: "gnu" isn't accurate unless using glibc

Comment (ararslan) [reply src/PlatformNames.jl]: Good call

**Comment by ararslan** [comment]
The build error on Julia 0.6 on macOS in the Travis PR build is unrelated. So I think this is ready whenever you are, @staticfloat.

**Comment by staticfloat** [comment]
Thanks Alex

**Comment by ararslan** [comment]
👌

**Pull request merged**